### PR TITLE
mssql_script: allow non-returning SQL statements

### LINKED ISSUE #6192: Module mssql_script using insert command returning failed on playbook but it's actually insert the records
### Summary

I've tried using the insert command and it's actually adding the record successfully but the playbook return failed, I've search for the python module pymssql which this module is using this python library, found that in python we can add conn.commit() and the error is gone but how can I do this in yaml playbook?

### Issue Type

Bug Report

### Component Name

mssql_script

### Ansible Version

```console (paste below)
$ ansible 2.9.6

```

### Community.general Version

```console (paste below)
$ ansible-galaxy collection list community.general.mssql_script

```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed

```
ansible 2.9.6
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.8.10 (default, Nov 14 2022, 12:59:47) [GCC 9.4.0]

### OS / Environment

Ubuntu 20.04

### Steps to Reproduce

```yaml (paste below)

```
- name: test command
  hosts: win
  gather_facts: false
  tasks:
    - name: DB command
      community.general.mssql_script:
        login_user: master-user
        login_password: *******
        login_host: "{{ inventory_hostname }}"
        login_port: 1433
        db: my_own_db
        #script: "select * from my_own_db.dbo.[my-table-1]"
        script: |
          INSERT INTO my_own_db.dbo.[my-table-1] (name, message) VALUES (%(name)s, %(message)s)
        params:
          name: myname2
          message: test message2
      register: output
      delegate_to: localhost

    - debug:
        msg:
        - "{{ output.query_results }}"

### Expected Results

Should return with green color instead of red

### Actual Results

```console (paste below)

```
`The full traceback is:
  File "/tmp/ansible_community.general.mssql_script_payload_1__m0256/ansible_community.general.mssql_script_payload.zip/ansible_collections/community/general/plugins/modules/mssql_script.py", line 282, in run_module
  File "src/pymssql/_pymssql.pyx", line 554, in pymssql._pymssql.Cursor.fetchall
fatal: [13.214.77.48 -> localhost]: FAILED! => {
    "changed": true,
    "error": "Statement not executed or executed statement has no resultset",
    "invocation": {
        "module_args": {
            "db": "nook_db",
            "login_host": "my-ip_address",
            "login_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "login_port": 1433,
            "login_user": "master-user",
            "name": "my_own_db",
            "output": "default",
            "params": {
                "message": "test message2",
                "name": "myname2"
            },
            "script": "INSERT INTO my_own_db.dbo.[my-table-1] (name, message) VALUES (%(name)s, %(message)s)\n"
        }
    },
    "msg": "query failed",
    "query": "INSERT INTO my_own_db.dbo.[my-table-1] (name, message) VALUES (%(name)s, %(message)s)\n`

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by ljkimmel**
- The current implementation fails out when certain statements or batches do not have resultsets - this limits the usefulness of the module
- Instead, it is known that statements without resultsets return then OperationalError exception with text "Statement not executed or executed statement has no resultset". We will utilize these facts to accept these statements
- The implementation also assumes that users will always use best- practices for the script syntax; that is, "GO" will always be capitalized but this is not strictly required -- update to allow "GO" to be any mixed-case
- Fixes #6192

### CODE REVIEW

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! As a first comment: can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment)? Thanks.

**Comment by ljkimmel** [comment]
@felixfontein I added the changelog fragment as requested. My apologies if it's not sufficient. This is one of my first Github PRs.

File: changelogs/fragments/6192-allow-empty-resultsets.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - mssql_script_module - Handle error condition for empty resultsets to allow for non-returning SQL statements (eg. UPDATE/INSERT)
+  - mssql_script_module - Allow for "GO" statement to be mixed-case for scripts not using strict syntax
```
Comment (felixfontein) [inline changelogs/fragments/6192-allow-empty-resultsets.yml:3]:
```suggestion
  - mssql_script - handle error condition for empty resultsets to allow for non-returning SQL statements (for example ``UPDATE`` and ``INSERT``) (https://github.com/ansible-collections/community.general/pull/6457).
  - mssql_script - allow for ``GO`` statement to be mixed-case for scripts not using strict syntax (https://github.com/ansible-collections/community.general/pull/6457).
```

Comment (ljkimmel) [reply changelogs/fragments/6192-allow-empty-resultsets.yml]: @felixfontein I have applied your suggestions as well as updated the batching logic to address some issues I discovered during testing with pulling (slurp) scripts from Windows machines. The logic is more complicated but, I believe, more reliably batches scripts from a wider array of sources.

File: changelogs/fragments/6192-allow-empty-resultsets.yml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+minor_changes:
+  - mssql_script - handle error condition for empty resultsets to allow for non-returning SQL statements (for example ``UPDATE`` and ``INSERT``) (https://github.com/ansible-collections/community.general/pull/6457).
+  - mssql_script - allow for ``GO`` statement to be mixed-case for scripts not using strict syntax (https://github.com/ansible-collections/community.general/pull/6457).
+  - mssql_script - improve batching logic to allow a wider variety of input scripts. For example, SQL scripts slurped from Windows machines which may contain carriage return (''\r'') characters.
```
Comment (felixfontein) [inline changelogs/fragments/6192-allow-empty-resultsets.yml:4]:
```suggestion
  - mssql_script - improve batching logic to allow a wider variety of input scripts. For example, SQL scripts slurped from Windows machines which may contain carriage return (''\r'') characters (https://github.com/ansible-collections/community.general/pull/6457).
```

**Review by russoz — CHANGES_REQUESTED** [review]

File: tests/integration/targets/mssql_script/tasks/main.yml [hunk 3]
```diff
@@ -49,6 +49,19 @@
     login_port: "{{ mssql_port }}"
     script: "SELECT 1"
 
+- name: Execute a malformed query
+  ws1cs.db.mssql_script:
```
[the remark this answers is not in the document]
Comment (ljkimmel) [reply tests/integration/targets/mssql_script/tasks/main.yml:53]: Yep. Good catch. We incorporated this into one of our own collections while waiting for this PR to be merged. I missed changing the FQCN back. Will fix.

File: plugins/modules/mssql_script.py [hunk 4]
```diff
@@ -280,23 +280,50 @@ def run_module():
...
+            # We know we executed the statement so this error just means we have no resultset
+            # which is ok (eg UPDATE/INSERT)
+            if (
+                type(e).__name__ == 'OperationalError' and
+                str(e) == 'Statement not executed or executed statement has no resultset'
+            ):
+                query_results.append([])
+            else:
+                error_msg = '%s: %s' % (type(e).__name__, str(e))
+                return module.fail_json(msg="query failed", query=query, error=error_msg, **result)
```
[the remark this answers is not in the document]
Comment (ljkimmel) [reply plugins/modules/mssql_script.py:326]: @russoz thanks for the feedback. I see what you mean but does that really clarify it enough to justify the need for redundant code (error_msg=..., return module.fail_json...)?

[the remark this answers is not in the document]
Comment (russoz) [reply plugins/modules/mssql_script.py:326]: _That_ redundancy is what I suggested solving by letting the exceptions propagate through the stack and handling with fail_json() in main() instead.

btw, it is not `return module.fail_json()` - there is no return from that call.

[the remark this answers is not in the document]
Comment (russoz) [reply plugins/modules/mssql_script.py:326]: On the question of being enough to justify, that sounds like a long and subjective discussion.

[the remark this answers is not in the document]
Comment (ljkimmel) [reply plugins/modules/mssql_script.py:326]: @russoz thanks again for the feedback. Another good catch on that `return`. That was in the original code and it worked so I didn't even question it and carried it forward.

[the remark this answers is not in the document]
Comment (ljkimmel) [reply plugins/modules/mssql_script.py]: @russoz I've implemented most of your suggestions. I started implementing the try/catch around the `run_module()` call and shortly realized that several of the items needed for the error we want to produce are created, and exist only, within the scope of the `run_module()` function. I feel like making these global variables just to have access to them outside of that scope starts to make things less clear. Therefore, for now, I'll leave as-is unless that's a show-stopper for you.

**Review by russoz — APPROVED** [review]
Fair enough. LGTM

**Comment by felixfontein** [comment]
@ljkimmel thanks for your contribution!
@russoz thanks for reviewing!

**Pull request merged**